### DISCUSSION TITLE
Credhub and conjur credential managers

### DISCUSSION THREAD

**Discussion by gaigaslab-operations** [concourse/concourse#6135] [Help & Support]
We have been using credhub as our credential manager for several years without issue.
We are adding conjur as an additional credential manager. credhub holds the ‘platform’ secrets required to integrate with the CI/CD environment and the platforms, while conjur will hold the ‘application’ secrets required to run the deployed applications.

My question(s) is about troubleshooting this dual-credential manager integration.
My credhub credentials are stored in /concourse//{{.secret}} and referenced using ((secret))
My conjur credentials are stored in ENV/GROUP/GROUP_{{.Team}}_D/delegation/consumers/{{.secret}} and referenced using ((conjur.secret))

Is ((conjur.secret)) the correct format? I see nowhere where I can actually name the conjur credential manager in the documentation

How can I troubleshoot the error:
failed to interpolate task config: undefined vars: conjur.pcf-testd2

**Comment by gaigaslab-operations** [comment]
Also, how can we set the global credential manager to credhub so that conjur is only used when prefixed?

When we enabled both conjur and the previously enabled credhub in our bosh deployment, all of our credhub secrets started failing.

**Comment by chenbh** [comment]
Concourse currently only supports a single global credential manager. Additional pipeline level cred managers can be configured as [var sources](https://concourse-ci.org/vars.html#var-sources), but those are still experimental and are limited to `vault` and `ssm` for now (other implementations are tracked in https://github.com/concourse/concourse/issues/5813). 

There's more details about var sources and how they're configured in the [var sources RFC](https://github.com/concourse/rfcs/blob/master/039-var-sources/proposal.md)

**Reply by gaigaslab-operations** [reply]
So how do you set the global credential manager?

**Reply by jamieklassen** [reply]
expanding a bit, even though it's possible to provide config for multiple credential managers, this loop has a `break` at the end of its body:

https://github.com/concourse/concourse/blob/95f7540e0e6b34c5db8ecef2b77108b50e8cc0c7/atc/atccmd/command.go#L1260-L1287

so whichever manager is configured "first" is the one that concourse will use. This is a bit awkward and if we explictly want there to be only one, maybe we should error on startup if multiple configs are provided.

**Reply by jamieklassen** [reply]
@gaigaslab-operations this may also be lacking in the docs, but the "global credential manager" is the one you configure by providing flags/envvars to the `concourse web` process.

**Reply by jamieklassen** [reply]
so the tl;dr of @chenbh's comment is that your use case isn't supported. Unless we add support for conjur as a var source, you can either use credhub or conjur, but not both.

**Reply by gaigaslab-operations** [reply]
Thank you @jamieklassen!

So, where does 'add[ing] support for conjur as a var source' rank in the to-do list?  The security folks are NOT going to be happy.

**Reply by jamieklassen** [reply]
😬 it's actually not on the roadmap, because the plan was to take a slightly longer view:

from https://concourse-ci.org/vars.html#var-sources:
> Currently, only two types are supported: `vault` and `dummy`. In the future we want to make use of something like the [Prototypes (RFC #37)](https://github.com/concourse/rfcs/pull/37) so that third-party credential managers can be used just like resource types.

and https://concourse-ci.org/vars.html#cluster-wide-credential-manager:
> In the future we would like to introduce support for multiple cluster-wide var sources, configured using the `var_source` schema, and begin deprecating the cluster-wide credential manager.

EDIT: as @chenbh mentioned - while undocumented, `ssm` is also supported as a `var_source`.

**Reply by gaigaslab-operations** [reply]
Any chance `conjur` is an undocumented, but supported `var_source`? :-)

**Reply by jamieklassen** [reply]
it's not, but the difference wouldn't be huge - adding `ssm` was done in #5576 - if you really had no other choice you might be able to PR the change and wait for the next release...

**Reply by jamieklassen** [reply]
there might be other long-shot solutions like having one "platform" concourse cluster that accesses credhub and a separate "app" concourse cluster that accesses conjur. Obviously this would only work if you never need access to both managers in the same pipeline.
